docs/design/2022-04-15-multi-schema-change.md

```diff
@@ -0,0 +1,137 @@
+# Proposal: Support Multi-Schema Change
+
+- Author(s): [tangenta](https://github.com/tangenta)
+- Tracking Issue: https://github.com/pingcap/tidb/issues/14766
+
+## Abstract
+
+This proposes an implementation of Multi-Schema Change.
+
+## Background
+
+Multi-Schema Change is one of MySQL's extended features to the SQL standard. It allows the users to atomically make multiple schema changes in one SQL statement, including column and index `ADD`, `ALTER`, `DROP`, and `CHANGE`, as well as table option changes. For example:
+
+```sql
+CREATE TABLE t (a INT, c INT);
+ALTER TABLE t ADD COLUMN b INT AUTO_INCREMENT PRIMARY KEY, 
+ MODIFY COLUMN c CHAR(5),
+ ADD INDEX idx(a),
+ AUTO_INCREMENT = 1000;
+```
+
+Currently, TiDB only supports one schema change per SQL statement and multi-schema changes in rare cases.
+
+When users try to migrate data from MySQL-like databases, they have to spend extra effort to rewrite a multi-schema change DDL to several single-schema change DDLs. For the users who rely on ORM frameworks like flyway to construct SQLs automatically, rewriting SQL is not even possible.
+
+Above all, Multi-Schema Change can be a blocking issue for those who want to use TiDB.
+
+## Proposal
+
+The implementation is based on the [online DDL architecture](https://github.com/pingcap/tidb/blob/e0c461a84cf4ad55c7b51c3f9db7f7b9ba51bb62/docs/design/2018-10-08-online-DDL.md). Similar to the DDL job, we introduce a new structure "sub-job", which represents a single schema change. As its name suggests, a job can contain zero or more sub-jobs.
+
+The Multi-Schema Change DDL jobs have the type `ActionMultiSchemaChange`. In the current worker model, there is a dedicated code path (`onMultiSchemaChange()`) to run these jobs. Only Multi-Schema Change jobs can have sub-jobs.
+
+For example, the DDL
+
+```SQL
+ALTER TABLE t ADD COLUMN b INT, MODIFY COLUMN a CHAR(255);
+```
+
+can be modeled as a job like
+
+```go
+job := &Job {
+    Type: ActionMultiSchemaChange,
+    MultiSchemaInfo: &MultiSchemaInfo {
+        SubJobs: []*SubJob {
+            &SubJob {
+                Type: ActionAddColumn,
+                Args: ...
+            },
+            &SubJob {
+                Type: ActionModifyColumn,
+                Args: ...
+            },
+        }
+    } 
+}
+```
+
+In this way, we pack multiple schema changes into one job. Just like any other jobs, it enters the DDL job queue stored in TiKV and waits for a suitable worker to pick it up.
```

Comment:
DDL job queue and the worker is the detail implementation of the `online DDL`, we may say something like `the job will be recorded and handled`.

Reply:
According to @bb7133 's suggestion, I changed it to "pick it up and handle it."